README.md
## geohash
A package for encoding, decoding and neighbouring geohashes

__Author:__ Oliver Keyes<br/>
__License:__ [MIT](http://opensource.org/licenses/MIT)<br/>
__Status:__ Stable

[![Travis-CI Build Status](https://travis-ci.org/Ironholds/geohash.svg?branch=master)](https://travis-ci.org/Ironholds/geohash) [![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/geohash)](https://cran.r-project.org/package=geohash) ![downloads](http://cranlogs.r-pkg.org/badges/grand-total/geohash)

### Description

*geohashes* are a way of representing latitude/longitude pairs as individual,
short strings of numbers and letters. The <code>geohash</code> package provides
tools for:

* Encoding latitude/longitude pairs into geohashes;
* Decoding geohashes into latitude/longitude pairs;
* Identifying the neighbouring geohashes of a provided geohash (or vector of geohashes).

For more information, see the [geohash vignette](https://github.com/Ironholds/geohash/blob/master/vignettes/geohash.Rmd).

Please note that this project is released with a [Contributor Code of Conduct](https://github.com/Ironholds/geohash/blob/master/CONDUCT.md).
By participating in this project you agree to abide by its terms.

### Installation
    
To get the development version (currently recommended):

    devtools::install_github("ironholds/geohash")
    
The latest CRAN version can be obtained via:

    install.packages("geohash")


NEWS
Version 0.4.0
-------------------------------------------------------------------------

* Output of `gh_decode` now includes a column `gh` reflecting the input in the order received, [#12](https://github.com/Ironholds/geohash/issues/12)). Thanks to @MichaelChirico for filing & completing the request.

Version 0.3.0
-------------------------------------------------------------------------

* The international dateline no longer snarls up hash generation, decoding and comparison (#7, thanks to Seth Wenchel for the report and fix)

* Operating systems running GCC 7.1 are now supported (#6)

* Function registration is integrated

Version 0.2.0
-------------------------------------------------------------------------

BUG FIXES
* gh_encode has a default precision of 6, errors on negative precision values, and falls back to 10 for >10 values. (#4)
* geohash now works under gcc 7.1 (#6)

Version 0.1.2
-------------------------------------------------------------------------
BUG FIXES
* Longitudes are now correctly returned from gh_decode (thanks to Alun Hewinson for spotting this) (#1)

Version 0.1.1
-------------------------------------------------------------------------

* Initial release


tests/testthat/test_decode.R
testthat::context("Test geohash decoding")

testthat::test_that("A simple lat/lng pair will correctly decoding", {

  result <- gh_decode("ezs42")
  testthat::expect_equal(result$gh[1L], structure(1L, .Label = "ezs42", class = "factor"))
  testthat::expect_lt(result$lat[1], 43)
  testthat::expect_gt(result$lat[1], 42)
  testthat::expect_lt(result$lng[1], -4)
  testthat::expect_gt(result$lng[1], -6)
})

testthat::test_that("NAs are appropriately handled when decoding", {

  result <- gh_decode(c("ezs42", NA))
  testthat::expect_true(is.na(result$lat[2]))

})


tests/testthat/test_encode.R
testthat::context("Test geohash encoding")

testthat::test_that("A simple lat/lng pair will correctly encode", {

  hash <- gh_encode(42.60498046875, -5.60302734375, 5)
  testthat::expect_equal(length(hash), 1)
  testthat::expect_equal(hash[1], "ezs42")

})

test_that("Incorrect lengths are appropriately handled when encoding", {

  testthat::expect_error(gh_encode(c(42.60498046875,12.3), -5.60302734375, 5),
                         regexp = "The lats and lngs vectors must be the same size")

})

testthat::test_that("NAs are appropriately handled when encoding", {

  hash <- gh_encode(c(12.3, NA), c(1.34,-5.60302734375), 5)

  testthat::expect_true(is.na(hash[2]))

})

testthat::test_that("an NA or <1 precision produces an error", {

  testthat::expect_error(gh_encode(12.3, 1.3, -1))
  testthat::expect_error(gh_encode(12.3, 1.3, NA_integer_))

})


testthat::test_that(">10 precision produces a warning and 10-length entry", {

  testthat::expect_warning(gh_encode(12.3, 1.3, 11))
  testthat::expect_equal(nchar(gh_encode(12.3, 1.3, 11)), 10)

})


src/cgeohash.h
#ifndef __CGEOHASH_HPP__
#define __CGEOHASH_HPP__

#include <Rcpp.h>
using namespace Rcpp;
namespace cgeohash {

struct DecodedBBox {
    double minlat, minlon, maxlat, maxlon;
};

struct DecodedHash {
    double latitude;
    double longitude;

    double latitude_err;
    double longitude_err;
};

// Encode a pair of latitude and longitude into geohash
std::string encode(double latitude, double longitude, unsigned int precision);

// Decode a hash string into pair of latitude and longitude
DecodedHash decode(std::string hash_string);

// Decode hashstring into a bound box matches it
DecodedBBox decode_bbox(std::string hash_string);

// Find neighbor of a geohash string in certain direction.
// Direction is a two-element array:
// Ex: [ 1, 0] == north
// Ex: [-1,-1] == southwest
std::string neighbor(std::string hash_string, const int direction []);

std::string neighbor(DecodedHash lonlat, const int direction [], const int length);

std::vector < std::string > all_neighbours(std::string hash);

double fix_longitude(double longitude);
}

#endif


R/RcppExports.R
# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#'@title Encode Latitude and Longitude Values
#'@description \code{\link{gh_encode}} generates geohashes of a
#'given precision from vectors of latitude and longitude values.
#'
#'@param lats a numeric vector of latitude values.
#'
#'@param lngs a numeric vector of longitude values. Must be the same length
#'as \code{lat}.
#'
#'@param precision an integer representing the precision the hashes should have.
#'This should be between 1 and 10; if the precision requested is greater than 10, it will
#'use 10 - if less than 1, it will error.
#'
#'@return a character vector of hashes, the same length as \code{lat} and \code{lng},
#'with \code{NA} values where one of the equivalent lat/lng pair was NA.
#'
#'@examples
#'# A simple example:
#'gh_encode(lat = 42.60498046875, lng = -5.60302734375, precision = 5)
#'#[1] "ezs42"
#'
#'@seealso \code{\link{gh_decode}}, for taking geohashes and turning them back into coordinates,
#'and \code{\link{gh_neighbours}} for retrieving the neighbouring hashes to a particular hash.
#'
#'@export
gh_encode <- function(lats, lngs, precision = 6L) {
    .Call(`_geohash_gh_encode`, lats, lngs, precision)
}

#'@title Decode Geohashes
#'@description \code{gh_decode} takes geohashes and turns them back into
#'latitude/longitude pairs, with an associated margin of error for each value.
#'
#'@param hashes a character vector of geohashes.
#'
#'@return a data.frame of five columns; "gh", "lat", "lng", "lat_error" and "lng_error"
#'
#'@seealso \code{\link{gh_encode}} for generating geohashes, and
#'\code{\link{gh_neighbours}} for identifying the neighbouring hash boxes
#'to a geohash.
#'
#'@examples
#'# A simple example:
#'gh_encode(lat = 42.60498046875, lng = -5.60302734375, precision = 5)
#'#[1] "ezs42"
#'
#'gh_decode("ezs42")
#'#      gh      lat       lng  lat_error  lng_error
#'# 1 ezs42 42.60498 -5.603027 0.02197266 0.02197266
#'@export
gh_decode <- function(hashes) {
    .Call(`_geohash_gh_decode`, hashes)
}

#'@rdname neighbours
#'@export
gh_neighbours <- function(hashes) {
    .Call(`_geohash_gh_neighbours`, hashes)
}

gh_neighbour <- function(hashes, direction) {
    .Call(`_geohash_gh_neighbour`, hashes, direction)
}



src/cgeohash.cpp
#include "cgeohash.h"
#include <math.h>

namespace cgeohash {

// Static array of 0-9, a-z
const char base32_codes[] = {
    '0',
    '1',
    '2',
    '3',
    '4',
    '5',
    '6',
    '7',
    '8',
    '9',
    'b',
    'c',
    'd',
    'e',
    'f',
    'g',
    'h',
    'j',
    'k',
    'm',
    'n',
    'p',
    'q',
    'r',
    's',
    't',
    'u',
    'v',
    'w',
    'x',
    'y',
    'z'
};

// direction constants
const int dir_east[] = {0,1};
const int dir_southeast[] = {-1,1};
const int dir_south[] = {-1,0};
const int dir_southwest[] = {-1,-1};
const int dir_west[] = {0,-1};
const int dir_northwest[] = {1,-1};
const int dir_north[] = {1,0};
const int dir_northeast[] = {1,1};

// Reverse map of characters --> index position
const std::map<char, int> build_base32_indexes(){
  std::map<char, int> output;

  for(unsigned int i = 0; i < strlen(base32_codes); i++) {
      output.insert( std::pair<char, int>(base32_codes[i], i) );
  }

  return output;
}

// Build a map of characters -> index position from the above array
const std::map<char, int> base32_indexes = build_base32_indexes();

// Convert the index position to the character in the array
char base32_codes_value_of(int index)
{
  return base32_codes[index];
}

// Convert the character to the index position in the array
int base32_codes_index_of(char c)
{
  return base32_indexes.find(c)->second;
}

std::string encode(double latitude, double longitude, unsigned int precision){
  // DecodedBBox for the lat/lon + errors
  std::string output;
  DecodedBBox bbox;
  bbox.maxlat = 90;
  bbox.maxlon = 180;
  bbox.minlat = -90;
  bbox.minlon = -180;
  double mid        = 0;
  bool   islon      = true;
  int    num_bits   = 0;
  int    hash_index = 0;

  // Pre-Allocate the hash string
  output = std::string(precision, ' ');
  unsigned int output_length = 0;

  while(output_length < precision) {
      if (islon) {
          mid = (bbox.maxlon + bbox.minlon) / 2;
          if(longitude > mid) {
              hash_index = (hash_index << 1) + 1;
              bbox.minlon=mid;
          } else {
              hash_index = (hash_index << 1) + 0;
              bbox.maxlon=mid;
          }
      } else {
          mid = (bbox.maxlat + bbox.minlat) / 2;
          if(latitude > mid ) {
              hash_index = (hash_index << 1) + 1;
              bbox.minlat = mid;
          } else {
              hash_index = (hash_index << 1) + 0;
              bbox.maxlat = mid;
          }
      }
      islon = !islon;

      ++num_bits;
      if (5 == num_bits) {
          // Append the character to the pre-allocated string
          // This gives us roughly a 2x speed boost
          output[output_length] = base32_codes[hash_index];

          output_length++;
          num_bits   = 0;
          hash_index = 0;
      }
  }
  return output;
}

DecodedBBox decode_bbox(std::string _hash_string){
    // Copy of the string down-cased
    // Wish this was ruby, then it would be simple: _hash_string.downcase();
    std::string hash_string(_hash_string);
    std::transform(
        _hash_string.begin(),
        _hash_string.end(),
        hash_string.begin(),
        ::tolower);

    DecodedBBox output;
    output.maxlat = 90;
    output.maxlon = 180;
    output.minlat = -90;
    output.minlon = -180;

    bool islon = true;

    for(int i = 0, max = hash_string.length(); i < max; i++) {
        int char_index = base32_codes_index_of(hash_string[i]);

        for (int bits = 4; bits >= 0; --bits) {
            int bit = (char_index >> bits) & 1;
            if (islon) {
                double mid = (output.maxlon + output.minlon) / 2;
                if(bit == 1) {
                    output.minlon = mid;
                } else {
                    output.maxlon = mid;
                }
            } else {
                double mid = (output.maxlat + output.minlat) / 2;
                if(bit == 1) {
                    output.minlat = mid;
                } else {
                    output.maxlat = mid;
                }
            }
            islon = !islon;
        }
    }
    return output;
}

DecodedHash decode(std::string hash_string){
    DecodedBBox bbox = decode_bbox(hash_string);
    DecodedHash output;
    output.latitude      = (bbox.minlat + bbox.maxlat) / 2;
    output.longitude     = (bbox.minlon + bbox.maxlon) / 2;
    output.latitude_err  = bbox.maxlat - output.latitude;
    output.longitude_err = bbox.maxlon - output.longitude;
    return output;
}

std::string neighbor(std::string hash_string, const int direction [])
{
  DecodedHash lonlat = decode(hash_string);

  return neighbor(lonlat, direction, hash_string.length());
}

std::string neighbor(DecodedHash lonlat, const int direction [], const int length)
{
  // Adjust the DecodedHash for the direction of the neighbors
  lonlat.latitude   += direction[0] * lonlat.latitude_err * 2;
  lonlat.longitude  += direction[1] * lonlat.longitude_err * 2;

  //pole check
  if(fabs(lonlat.latitude)>90.){
    lonlat.latitude   -= direction[0] * lonlat.latitude_err * 2;
    lonlat.longitude  += 180.0;
  }

  return encode(lonlat.latitude, fix_longitude(lonlat.longitude), length);
}

std::vector < std::string > all_neighbours(std::string hash){

  std::vector < std::string > output(8);
  DecodedHash lonlat = decode(hash);
  unsigned int input_size = hash.size();

  output[0] = neighbor(lonlat, dir_north, input_size); // North
  output[1] = neighbor(lonlat, dir_northeast, input_size); // Northeast
  output[2] = neighbor(lonlat, dir_east, input_size); // East
  output[3] = neighbor(lonlat, dir_southeast, input_size); // Southeast
  output[4] = neighbor(lonlat, dir_south, input_size); // South
  output[5] = neighbor(lonlat, dir_southwest, input_size); // Southwest
  output[6] = neighbor(lonlat, dir_west, input_size); // West
  output[7] = neighbor(lonlat, dir_northwest, input_size); // Northwest
  return output;
}


double fix_longitude(double longitude){
  if(longitude < -180.){
    longitude += 360.;
    return longitude;
  }
  if(longitude > 180.){
    longitude -= 360.;
    return longitude;
  }
  return(longitude);
}
} // end namespace cgeohash


src/RcppExports.cpp
// Generated by using Rcpp::compileAttributes() -> do not edit by hand
// Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#include <Rcpp.h>

using namespace Rcpp;

// gh_encode
CharacterVector gh_encode(NumericVector lats, NumericVector lngs, IntegerVector precision);
RcppExport SEXP _geohash_gh_encode(SEXP latsSEXP, SEXP lngsSEXP, SEXP precisionSEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< NumericVector >::type lats(latsSEXP);
    Rcpp::traits::input_parameter< NumericVector >::type lngs(lngsSEXP);
    Rcpp::traits::input_parameter< IntegerVector >::type precision(precisionSEXP);
    rcpp_result_gen = Rcpp::wrap(gh_encode(lats, lngs, precision));
    return rcpp_result_gen;
END_RCPP
}
// gh_decode
DataFrame gh_decode(CharacterVector hashes);
RcppExport SEXP _geohash_gh_decode(SEXP hashesSEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< CharacterVector >::type hashes(hashesSEXP);
    rcpp_result_gen = Rcpp::wrap(gh_decode(hashes));
    return rcpp_result_gen;
END_RCPP
}
// gh_neighbours
DataFrame gh_neighbours(CharacterVector hashes);
RcppExport SEXP _geohash_gh_neighbours(SEXP hashesSEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< CharacterVector >::type hashes(hashesSEXP);
    rcpp_result_gen = Rcpp::wrap(gh_neighbours(hashes));
    return rcpp_result_gen;
END_RCPP
}
// gh_neighbour
CharacterVector gh_neighbour(CharacterVector hashes, IntegerVector direction);
RcppExport SEXP _geohash_gh_neighbour(SEXP hashesSEXP, SEXP directionSEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< CharacterVector >::type hashes(hashesSEXP);
    Rcpp::traits::input_parameter< IntegerVector >::type direction(directionSEXP);
    rcpp_result_gen = Rcpp::wrap(gh_neighbour(hashes, direction));
    return rcpp_result_gen;
END_RCPP
}

static const R_CallMethodDef CallEntries[] = {
    {"_geohash_gh_encode", (DL_FUNC) &_geohash_gh_encode, 3},
    {"_geohash_gh_decode", (DL_FUNC) &_geohash_gh_decode, 1},
    {"_geohash_gh_neighbours", (DL_FUNC) &_geohash_gh_neighbours, 1},
    {"_geohash_gh_neighbour", (DL_FUNC) &_geohash_gh_neighbour, 2},
    {NULL, NULL, 0}
};

RcppExport void R_init_geohash(DllInfo *dll) {
    R_registerRoutines(dll, NULL, CallEntries, NULL, NULL);
    R_useDynamicSymbols(dll, FALSE);
}


src/geohash.cpp
#include "cgeohash.h"

//'@title Encode Latitude and Longitude Values
//'@description \code{\link{gh_encode}} generates geohashes of a
//'given precision from vectors of latitude and longitude values.
//'
//'@param lats a numeric vector of latitude values.
//'
//'@param lngs a numeric vector of longitude values. Must be the same length
//'as \code{lat}.
//'
//'@param precision an integer representing the precision the hashes should have.
//'This should be between 1 and 10; if the precision requested is greater than 10, it will
//'use 10 - if less than 1, it will error.
//'
//'@return a character vector of hashes, the same length as \code{lat} and \code{lng},
//'with \code{NA} values where one of the equivalent lat/lng pair was NA.
//'
//'@examples
//'# A simple example:
//'gh_encode(lat = 42.60498046875, lng = -5.60302734375, precision = 5)
//'#[1] "ezs42"
//'
//'@seealso \code{\link{gh_decode}}, for taking geohashes and turning them back into coordinates,
//'and \code{\link{gh_neighbours}} for retrieving the neighbouring hashes to a particular hash.
//'
//'@export
//[[Rcpp::export]]
CharacterVector gh_encode(NumericVector lats, NumericVector lngs,
                          IntegerVector precision = 6){

  // Initial checks
  if(precision.size() < 1 || IntegerVector::is_na(precision[0])){
    Rcpp::stop("A value for precision must be provided, and cannot be NA");
  }
  if(precision[0] < 1){
    Rcpp::stop("Precision must be greater than 1");
  }

  signed int& precision_ref = precision[0];
  if(precision_ref > 10){
    Rcpp::warning("Precision is greater than 10, the maximum. 10 used instead.");
    precision_ref = 10;
  }

  const unsigned int input_size = lats.size();
  if(input_size != lngs.size()){
    Rcpp::stop("The lats and lngs vectors must be the same size");
  }

  // Set up output and run
  CharacterVector output(input_size);

  for(unsigned int i = 0; i < input_size; i++){

    //Check to see if the user has asked us to stop doing the thing.
    if((i % 10000) == 0){
      Rcpp::checkUserInterrupt();
    }

    // Check to see if one of the inputs is NA
    if(NumericVector::is_na(lats[i]) || NumericVector::is_na(lngs[i])){
      output[i] = NA_STRING;
    } else {
      // Do the thing.
      output[i] = cgeohash::encode(lats[i], lngs[i], precision_ref);
    }
  }

  return output;
}

//'@title Decode Geohashes
//'@description \code{gh_decode} takes geohashes and turns them back into
//'latitude/longitude pairs, with an associated margin of error for each value.
//'
//'@param hashes a character vector of geohashes.
//'
//'@return a data.frame of four columns; "lat", "lng", "lat_error" and "lng_error"
//'
//'@seealso \code{\link{gh_encode}} for generating geohashes, and
//'\code{\link{gh_neighbours}} for identifying the neighbouring hash boxes
//'to a geohash.
//'
//'@examples
//'# A simple example:
//'gh_encode(lat = 42.60498046875, lng = -5.60302734375, precision = 5)
//'#[1] "ezs42"
//'
//'gh_decode("ezs42")
//'# lat      lng      lat_error  lng_error
//'# 42.60498 -5.603027 0.02197266 0.02197266
//'@export
//[[Rcpp::export]]
DataFrame gh_decode(CharacterVector hashes){

  unsigned int input_size = hashes.size();
  NumericVector lats(input_size);
  NumericVector lngs(input_size);
  NumericVector lat_error(input_size);
  NumericVector lng_error(input_size);
  for(unsigned int i = 0; i < input_size; i++){

    if((i % 10000) == 0){
      Rcpp::checkUserInterrupt();
    }

    if(CharacterVector::is_na(hashes[i])){
      lats[i] = NA_REAL;
      lngs[i] = NA_REAL;
      lat_error[i] = NA_REAL;
      lng_error[i] = NA_REAL;
    } else {
      cgeohash::DecodedHash holding = cgeohash::decode(Rcpp::as<std::string>(hashes[i]));

      lats[i] = holding.latitude;
      lngs[i] = holding.longitude;
      lat_error[i] = holding.latitude_err;
      lng_error[i] = holding.longitude_err;
    }

  }

  return DataFrame::create(_["gh"] = hashes,
                           _["lat"] = lats,
                           _["lng"] = lngs,
                           _["lat_error"] = lat_error,
                           _["lng_error"] = lng_error,
                           _["stringsAsFactors"] = false);
}

//'@rdname neighbours
//'@export
// [[Rcpp::export]]
DataFrame gh_neighbours(CharacterVector hashes){

  unsigned int input_size = hashes.size();
  CharacterVector north(input_size);
  CharacterVector ne(input_size);
  CharacterVector east(input_size);
  CharacterVector se(input_size);
  CharacterVector south(input_size);
  CharacterVector sw(input_size);
  CharacterVector west(input_size);
  CharacterVector nw(input_size);
  std::vector < std::string > holding(8);

  for(unsigned int i = 0; i < input_size; i++){

    if((i % 10000) == 0){
      Rcpp::checkUserInterrupt();
    }

    if(CharacterVector::is_na(hashes[i])){
      north[i] = NA_REAL;
      ne[i] = NA_REAL;
      east[i] = NA_REAL;
      se[i] = NA_REAL;
      south[i] = NA_REAL;
      sw[i] = NA_REAL;
      west[i] = NA_REAL;
      nw[i] = NA_REAL;
    } else {

      holding = cgeohash::all_neighbours(Rcpp::as<std::string>(hashes[i]));

      north[i] = holding[0];
      ne[i] =  holding[1];
      east[i] =  holding[2];
      se[i] =  holding[3];
      south[i] =  holding[4];
      sw[i] =  holding[5];
      west[i] =  holding[6];
      nw[i] =  holding[7];
    }

  }

  return DataFrame::create(_["north"] = north,
                           _["northeast"] = ne,
                           _["east"] = east,
                           _["southeast"] = se,
                           _["south"] = south,
                           _["southwest"] = sw,
                           _["west"] = west,
                           _["northwest"] = nw,
                           _["stringsAsFactors"] = false);

}

// [[Rcpp::export]]
CharacterVector gh_neighbour(CharacterVector hashes, IntegerVector direction){

  unsigned int input_size = hashes.size();
  int directions [2] = {direction[0], direction[1]};
  CharacterVector output(input_size);
  for(unsigned int i = 0; i < input_size; i++){

    if((i % 10000) == 0){
      Rcpp::checkUserInterrupt();
    }

    if(CharacterVector::is_na(hashes[i])){
      output[i] = NA_STRING;
    } else {
      output[i] = cgeohash::neighbor(Rcpp::as<std::string>(hashes[i]), directions);
    }
  }

  return output;
}
